File: packages/aegis-ledger-sdk/src/replay.ts
import { canonicalize, sha256Hex } from "./canonical.js";
import { readEvents } from "./store.js";
import type { ApsEvent, ReplayResult, ReplayStep } from "./types.js";

const TIER_A_ACTIONS = new Set(["decision.record"]);

function reconstructTierA(event: ApsEvent): ReplayStep["reconstructed"] {
  const ctx = event.context;
  return {
    inputs: structuredClone(ctx.inputs ?? {}),
    model: ctx.model as Record<string, unknown> | undefined,
    policy: ctx.policy as Record<string, unknown> | undefined,
    outcome: structuredClone(ctx.outcome ?? {}),
  };
}

/**
 * Deterministic Tier-A replay: same ledger bytes → same digest.
 */
export function replay(storePath: string, traceId = "local"): ReplayResult {
  const events = readEvents(storePath);
  const steps: ReplayStep[] = [];

  for (const event of events) {
    if (!TIER_A_ACTIONS.has(event.action)) {
      continue;
    }
    steps.push({
      event_id: event.event_id,
      action: event.action,
      subject: event.subject,
      tier: "A",
      reconstructed: reconstructTierA(event),
    });
  }

  const digest = `sha256:${sha256Hex(canonicalize({ trace_id: traceId, steps }))}`;
  return { trace_id: traceId, steps, digest };
}

File: packages/aegis-sdk-py/salanor_aegis_ledger/replay.py
from __future__ import annotations

import hashlib
from pathlib import Path
from typing import Any

from salanor_aegis_ledger.canonical import canonicalize
from salanor_aegis_ledger.store import read_events

TIER_A_ACTIONS = {"decision.record"}


def replay(store_path: Path | str, trace_id: str = "local") -> dict[str, Any]:
    events = read_events(Path(store_path))
    steps: list[dict[str, Any]] = []
    for event in events:
        if event.get("action") not in TIER_A_ACTIONS:
            continue
        ctx = event["context"]
        steps.append(
            {
                "event_id": event["event_id"],
                "action": event["action"],
                "subject": event["subject"],
                "tier": "A",
                "reconstructed": {
                    "inputs": ctx.get("inputs", {}),
                    "model": ctx.get("model"),
                    "policy": ctx.get("policy"),
                    "outcome": ctx.get("outcome", {}),
                },
            }
        )
    digest = hashlib.sha256(canonicalize({"trace_id": trace_id, "steps": steps}).encode()).hexdigest()
    return {"trace_id": trace_id, "steps": steps, "digest": f"sha256:{digest}"}
